### PR TITLE
TC-FAN-3.1: Enhanced attribute value testing

### PR DESCRIPTION
**Pull request by raul-marquez-csa**
### Testing
Addresses **TC-FAN-3.1** from issue
[[TC-FAN] Fan tests do not sufficiently test attribute interactions #4788
](https://github.com/CHIP-Specifications/chip-test-plans/issues/4788)

_- Should step through all the modes available from the enum and ensure the percent values increase and decrease correctly with respect to the mode (ie, off is 0, high should be higher than medium, which should be higher than low, which should be non-zero as per the percent rules in the spec._

### Enhancement testing update
Four scenarios are tested

**Scenario 1**
  - Attribute to update: `PercentSetting`
  - Attribute to verify: `PercentSetting`, `PercentCurrent`, `FanMode`, and if present, `SpeedSetting` and `SpeedCurrent`
  - Update order: `Ascending`

**Scenario 2**
  - Attribute to update: `PercentSetting`
  - Attribute to verify: `PercentSetting`, `PercentCurrent`, `FanMode`, and if present, `SpeedSetting` and `SpeedCurrent`
  - Update order: `Descending`

**Scenario 3**
  - Attribute to update: `FanMode`
  - Attribute to verify: `PercentSetting`, `PercentCurrent`, `FanMode`, and if present, `SpeedSetting` and `SpeedCurrent`
  - Update order: `Ascending`

**Scenario 4**
  - Attribute to update: `FanMode`
  - Attribute to verify: `PercentSetting`, `PercentCurrent`, `FanMode`, and if present, `SpeedSetting` and `SpeedCurrent`
  - Update order: `Descending`

- [Fan Control spec](https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/app_clusters/FanControl.adoc#fan-control-cluster)

- [Corresponding test plan PR](https://github.com/CHIP-Specifications/chip-test-plans/pull/4965) - [Preview](https://github.com/raul-marquez-csa/chip-test-plans/blob/TC-FAN-3.1-update/src/cluster/FanControl.adoc#316-tc-fan-31-mandatory-functionality-with-dut_server)

### CODE REVIEW

File: src/python_testing/TC_FAN_3_1.py [hunk 1]
```diff
@@ -36,16 +36,43 @@
...
 import time
+import random
 
 import chip.clusters as Clusters
 from chip.interaction_model import Status
-from chip.testing.matter_testing import MatterBaseTest, async_test_body, default_matter_test_main
+from chip.testing.matter_testing import MatterBaseTest, TestStep, async_test_body, default_matter_test_main
 from mobly import asserts
 
+# import pdb
```
Comment (andy31415) [inline src/python_testing/TC_FAN_3_1.py:46]: delete commented out code

File: src/python_testing/TC_FAN_3_1.py [hunk 2]
```diff
@@ -73,97 +109,366 @@ def pics_TC_FAN_3_1(self) -> list[str]:
 
     @async_test_body
     async def test_TC_FAN_3_1(self):
+        # pdb.set_trace()
```
Comment (andy31415) [inline src/python_testing/TC_FAN_3_1.py:112]: same here: no commented out code.

File: src/python_testing/TC_FAN_3_1.py [hunk 3]
```diff
@@ -73,97 +109,366 @@ def pics_TC_FAN_3_1(self) -> list[str]:
...
+        # | 0 (Off)    | 0              | 0                    |
+        # +------------+----------------+----------------------+
+        # | 1 (Low)    | 1 - 30         | 10                   |
+        # +------------+----------------+----------------------+
+        # | 2 (Medium) | 31 - 70        | 40                   |
+        # +------------+----------------+----------------------+
+        # | 3 (High)   | 71 - 100       | 80                   |
+        # +------------+----------------+----------------------+
+        percent_setting_range_off = 0
+        percent_setting_range_low = random.randint(1, 30)
```
Comment (andy31415) [inline src/python_testing/TC_FAN_3_1.py:132]: I am a bit worried here as test is not reproducible/re-runnable and it may behave as flaky: if some random value ends up being "bad" (probably some edge case) a re-run of the test will make the test pass.

Could we:
  a) ensure we run the test on all edge cases (like 1, 30, 31, 70, 71, 100)
  b) for random (if the test plan requests it) set a fixed seed to ensure the test is identical in any case (even though we are using a random module without a random value ... so that is not that great either, but at least it is a stable test)

File: src/python_testing/TC_FAN_3_1.py [hunk 4]
```diff
@@ -36,16 +36,43 @@
...
+
 logger = logging.getLogger(__name__)
 
 
 class TC_FAN_3_1(MatterBaseTest):
+    
+    def steps_TC_FAN_3_1(self):
+        return [TestStep("1", "Commissioning already done."),
+                TestStep("2", "Action",
+                         "Verification"),
```
Comment (andy31415) [inline src/python_testing/TC_FAN_3_1.py:56]: These tests steps should have some better description' Action/Verification is not very useful.

File: src/python_testing/TC_FAN_3_1.py (hunk 3 above)
Comment (raul-marquez-csa) [reply src/python_testing/TC_FAN_3_1.py]: This will be fixed once out of draft PR

File: src/python_testing/TC_FAN_3_1.py (hunk 2 above)
Comment (raul-marquez-csa) [reply src/python_testing/TC_FAN_3_1.py]: This will be fixed once out of draft PR

File: src/python_testing/TC_FAN_3_1.py (hunk 1 above)
Comment (raul-marquez-csa) [reply src/python_testing/TC_FAN_3_1.py]: This will be fixed once out of draft PR

File: src/python_testing/TC_FAN_3_1.py [hunk 5]
```diff
@@ -34,136 +34,333 @@
...
+
+        # Logging the scenario being tested
+        await self.log_scenario(endpoint, attr_to_update, attr_to_verify, speed_setting_read, order)
+
+        # Write to attribute iteratively
+        attr_value_current = attr_value_read
+        attr_value_previous = attr_value_read
+        speed_setting_current = speed_setting_read
+        speed_setting_previous = speed_setting_read
+        for value_to_write in iteration_range:
```
Comment (cecille) [inline src/python_testing/TC_FAN_3_1.py:268]: I see what you're trying to do here, but I think the conditions here might be sufficiently different that your code is longer and more confusing by trying to make these the same and pipe them through a single function

for percent, you want:
- update percent
- check for percent update every time (report + read)
- if mode change happened, check that the mode is incrementing or decrementing as expected (report + read)
- ditto for speed
- at the end ensure the last mode change happened
- ditto for speed
- for both mode and speed, you want to make sure you saw all the possibilities (except auto) and you want to make sure you didn't see auto.

for update mode you want:
- update mode
- check that the mode change happened (report and read)
- check that the percent change happened (every time) and is going in the correct direction (report and read)
- speed change should happen every time IF the number of speed settings is larger than the number of modes

There's commonalities, but it might make more sense to break this as
- update attribute, wait for report (with INVALID_IN_STATE handing) as a single function
- if INVALID_IN_STATE, verify no changes
- if SUCCESS verify dependent attributes (reports and reads, with a parameter saying if the change is required)

File: src/python_testing/TC_FAN_3_1.py [hunk 6]
```diff
@@ -34,136 +34,333 @@
...
+        # Initializatization of the attribute to update (Write and read back verification)
+        await self.write_and_verify_attribute(endpoint, attr_to_update, value_init_update)
+
+        # Initializatization of the attribute to verify (Write and read back verification)
+        attr_value_read = await self.write_and_verify_attribute(endpoint, attr_to_verify, value_init_verify)
+
+        # Initializatization of the SpeedSetting attribute, if supported (Write and read back verification)
+        if self.supports_speed:
+            speed_setting_init = 0 if order == OrderEnum.Ascending else self.speed_max
+            speed_setting_read = await self.write_and_verify_attribute(endpoint, speed_setting_attr, speed_setting_init)
```
Comment (cecille) [inline src/python_testing/TC_FAN_3_1.py:254]: same thing here - speed should update itself when the initial attribute is updated.

File: src/python_testing/TC_FAN_3_1.py [hunk 7]
```diff
@@ -34,136 +34,333 @@
...
+            asserts.assert_less(value_current, value_previous,
+                                f"[FC] Current {attr_to_verify.__name__} must be less than previous {attr_to_verify.__name__}")
+
+        # Logging attribute value change details
+        sub_text = f"({value_previous}) to ({value_current})"
+        if attr_to_verify == Clusters.FanControl.Attributes.FanMode:
+            sub_text = f"({value_previous}:{value_previous.name}) to ({value_current}:{value_current.name})"
+        logging.info(f"\t\t[FC] {attr_to_verify.__name__} changed from {sub_text}")
+
+    async def log_scenario(self, endpoint, attr_to_update, attr_to_verify, speed_setting_read, order):
```
[the remark this answers is not in the document]
Comment (raul-marquez-csa) [reply src/python_testing/TC_FAN_3_1.py]: I updated log_scenario to not read the attributes, all attribute values are passed in now.

File: src/python_testing/TC_FAN_3_1.py [hunk 8]
```diff
@@ -34,136 +34,333 @@
...
+        # There are only 6 valid FanModeSequence values (0, 1, 2, 3, 4, 5)
+        # Verifying read FanModeSequence attribute value is valid
+        asserts.assert_in(fan_mode_sequence, range(0, 5), f"Unsupported FanModeSequence: {fan_mode_sequence}")
+
+        fan_mode_sequence_enum_value = self.get_enum_value(fan_mode_sequence)
+        logging.info(f"[FC] Supported fan modes: {fan_mode_sequence_enum_value.name}")
+
+        fan_modes = None
+        if fan_mode_sequence == 0:
+            fan_modes = [fm_en.kOff.value, fm_en.kLow.value, fm_en.kMedium.value, fm_en.kHigh.value]
```
[the remark this answers is not in the document]
Comment (raul-marquez-csa) [reply src/python_testing/TC_FAN_3_1.py]: Correct, updated to use enums as is

File: src/python_testing/TC_FAN_3_1.py [hunk 9]
```diff
@@ -34,136 +34,333 @@
...
+        start_time = time.time()
+        while time.time() - start_time < timeout_sec:
+            for q in list(queue):
+                if q.attribute == attribute:
+                    return q.value
+            await asyncio.sleep(0.01)
+        return None
+
+    @staticmethod
+    def get_enum_value(value) -> int:
```
[the remark this answers is not in the document]
Comment (raul-marquez-csa) [reply src/python_testing/TC_FAN_3_1.py]: Correct, removed get_enum_value function

File: src/python_testing/TC_FAN_3_1.py [hunk 10]
```diff
@@ -34,136 +34,333 @@
...
+                TestStep(12, "[FC] TH updates the value of the FanMode attribute of the DUT iteratively, in descending order (from 3 (High) to 0 (Off)) For each iteration, the TH reads the value of the PercentSetting attribute (and SpeedSetting, if supported) from the DUT.",
+                         "For each update, the DUT shall return either a SUCCESS or an INVALID_IN_STATE status code. Verify the following for the PercentSetting attribute value (and SpeedSetting, if supported): - For iterations where the DUT returns a SUCCESS status code after the FanMode attribute value update, and a change in the attribute value is detected, verify that the current value is less than the previous one. - For iterations where the DUT returns an INVALID_IN_STATE status code after the FanMode attribute value update, which means that no update operation occurred, verify that the current value is the same as the previous one."),
+                ]
+
+    @staticmethod
+    async def get_attribute_value_from_queue(queue, attribute, timeout_sec: float) -> Any:
```
[the remark this answers is not in the document]
Comment (raul-marquez-csa) [reply src/python_testing/TC_FAN_3_1.py:90]: I don't think **await_all_final_values_reported** can help me in this scenario.
Let's take the updating `PercentSetting` and monitoring `FanMode` scenario.

- We have no way of knowing before hand which `PercentSetting` value will trigger a `FanMode` change
- This is run iteratively, so not all `PercentSetting` written values will trigger a `FanMode` change
- Once we detect a `FanMode` change, we need the attribute value for verifying its progression (ascending/descending), and the **await_all_final_values_reported** function doesn't return anything

```
expected_final_value = [
    AttributeValue(
    endpoint,
    attribute=Clusters.FanControl.Attributes.FanMode,
    value=???????
    )]
self.attribute_subscription.await_all_final_values_reported(
    expected_final_value,
    timeout_sec=1)
```
So:

- Expected value? Unknown
- When to expect it? Unknown

I consider I need to keep the code as is, unless I'm missing something.

[the remark this answers is not in the document]
Comment (cecille) [reply src/python_testing/TC_FAN_3_1.py:90]: OK, I see what you're saying, but you're racing the flush here, which is why I'm a bit concerned.

What about if you used multiple subscriptions - one for the attribute being updated, so you can see it going in real-time, one for the whole cluster? In each iteration of the for loop, await the one being updated on the single attribute subscription, then at the end of the cycle, await the final sequence of all the expeced attribute reports. Then you can look back in the queue and make sure everything is normal - all modes separated by at least one percent updadte, modes in order, modes all there etc.

thoughts?

File: src/python_testing/TC_FAN_3_1.py [hunk 11]
```diff
@@ -34,136 +34,333 @@
...
+
+        # Get initialization parameters
+        attr_to_verify, value_init_verify, value_init_update, iteration_range = \
+            await self.get_initialization_parametes(attr_to_update, order)
+
+        # Initializatization of the attribute to update (Write and read back verification)
+        await self.write_and_verify_attribute(endpoint, attr_to_update, value_init_update)
+
+        # Initializatization of the attribute to verify (Write and read back verification)
+        attr_value_read = await self.write_and_verify_attribute(endpoint, attr_to_verify, value_init_verify)
```
[the remark this answers is not in the document]
Comment (raul-marquez-csa) [reply src/python_testing/TC_FAN_3_1.py]: This part of the script has been omitted now, see comment from above conversation.

File: src/python_testing/TC_FAN_3_1.py (hunk 5 above)
Comment (raul-marquez-csa) [reply src/python_testing/TC_FAN_3_1.py:277]: This has been updated and those variables have been set to None.
It's part of the change from this conversation:
[If you get INVALID_IN_STATE here, what happens?](https://github.com/project-chip/connectedhomeip/pull/36971#discussion_r1968225567)

This is in response to the first point of this comment.

I'm reviewing the rest of the comment.

File: src/python_testing/TC_FAN_3_1.py (hunk 6 above)
Comment (raul-marquez-csa) [reply src/python_testing/TC_FAN_3_1.py]: This is no longer ocurring

File: src/python_testing/TC_FAN_3_1.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (raul-marquez-csa) [reply src/python_testing/TC_FAN_3_1.py]: As per our previous conversation, I adapted the detection/verification mechanism to avoid the issues raised in your comment.

File: src/python_testing/TC_FAN_3_1.py (hunk 5 above)
Comment (raul-marquez-csa) [reply src/python_testing/TC_FAN_3_1.py:436]: As per our previous conversation, I adapted the detection/verification mechanism to avoid the issues raised in your comment.

File: src/python_testing/TC_FAN_3_1.py [hunk 12]
```diff
@@ -34,136 +34,333 @@
...
+        await self.log_scenario(endpoint, attr_to_update, attr_to_verify, speed_setting_read, order)
+
+        # Write to attribute iteratively
+        attr_value_current = attr_value_read
+        attr_value_previous = attr_value_read
+        speed_setting_current = speed_setting_read
+        speed_setting_previous = speed_setting_read
+        for value_to_write in iteration_range:
+            # Clear the attribute report queue before each update to avoid duplicates
+            self.attribute_subscription.get_last_report()
```
[the remark this answers is not in the document]
Comment (raul-marquez-csa) [reply src/python_testing/TC_FAN_3_1.py]: As per our previous conversation, I adapted the detection/verification mechanism to avoid the issues raised in your comment.

File: src/python_testing/TC_FAN_3_1.py [hunk 13]
```diff
@@ -34,136 +34,333 @@
...
+            logging.info(f"[FC] {attr_to_update.__name__} written: {value_to_write}")
+
+            # If the write status is SUCCESS, it means the write operation occurred
+            # Verify that the current attribute value is greater than the previous
+            # one if order was set to Ascending, or less if order was set to Descending
+            if write_status == Status.Success:
+                # Get current attribute value and verify correct increment/decrement
+                queue = self.attribute_subscription.attribute_queue.queue
+                attr_value_current = await self.get_attribute_value_from_queue(queue, attr_to_verify, timeout_sec)
+                if attr_value_current is not None:
```
[the remark this answers is not in the document]
Comment (raul-marquez-csa) [reply src/python_testing/TC_FAN_3_1.py]: As per our previous conversation, I adapted the detection/verification mechanism to avoid the issues raised in your comment.

**Review by cecille — COMMENTED** [review]
This code needs some restructuring to be readable. You have functions here with really large numbers of input arguments, and functions that appear to be doing two different things based on the input arguments, but checking them at every iteration.

This is just generally really hard to follow because things are grouped together in strange ways.

File: src/python_testing/TC_FAN_3_1.py [hunk 14]
```diff
@@ -30,140 +30,631 @@
...
+        elif self.fan_mode_sequence == 2:
+            fan_modes = [fm_enum.kOff, fm_enum.kLow, fm_enum.kMedium, fm_enum.kHigh, fm_enum.kAuto]
+        elif self.fan_mode_sequence == 3:
+            fan_modes = [fm_enum.kOff, fm_enum.kLow, fm_enum.kHigh, fm_enum.kAuto]
+        elif self.fan_mode_sequence == 4:
+            fan_modes = [fm_enum.kOff, fm_enum.kHigh, fm_enum.kAuto]
+        elif self.fan_mode_sequence == 5:
+            fan_modes = [fm_enum.kOff, fm_enum.kHigh]
+
+        if max_high:
```
[the remark this answers is not in the document]
Comment (raul-marquez-csa) [reply src/python_testing/TC_FAN_3_1.py:223]: I intend to re-use it.
I updated it to as per this comment.

File: src/python_testing/TC_FAN_3_1.py [hunk 15]
```diff
@@ -30,140 +30,631 @@
...
+            attribute (Any): The attribute to be written.
+            value (Any): The value to write to the attribute.
+
+        Returns:
+            Status: The status of the write operation.
+
+        Raises:
+            AssertionError: If the value read back does not match the value written.
+        """
+        write_status = await self.write_setting(endpoint, attribute, value)
```
[the remark this answers is not in the document]
Comment (raul-marquez-csa) [reply src/python_testing/TC_FAN_3_1.py:135]: Updated as per this comment.

File: src/python_testing/TC_FAN_3_1.py [hunk 16]
```diff
@@ -30,140 +30,631 @@
...
+
+    async def verify_attribute_success(self, attr_to_update, attr_to_verify, order, init_fan_mode, init_percent_setting, init_speed_setting, init_attr_to_verify_value, attr_to_verify_value_current, attr_to_verify_value_previous, iteration) -> tuple[Any, Any]:
+        """
+        Verifies the correct progression of the current and previous values of the attribute to verify.
+
+        Args:
+            attr_to_update (str): The attribute that was updated.
+            attr_to_verify (str): The attribute whose value needs to be verified.
+            order (str): The order in which the attributes are processed.
+            init_fan_mode (int): The initial fan mode setting.
```
[the remark this answers is not in the document]
Comment (raul-marquez-csa) [reply src/python_testing/TC_FAN_3_1.py:524]: It's not logging every iteration, it's logging the failed assertion message on any one attribute.
More than anything it's meant to provide sufficient detail on what went wrong so the user can quickly catch the error and provide a solution.

This function is run only when INVALID_IN_STATE, which in general is the exception, so it's not triggered on every iteration from 0 to 100.

Is your observation more related with esthetics? or performance?

File: src/python_testing/TC_FAN_3_1.py [hunk 17]
```diff
@@ -30,140 +30,631 @@
...
+        logging.info(f"\t\t[FC] {attr_to_verify.__name__} changed from {sub_text}")
+
+        # The fan's general state is Off
+        if init_fan_mode == fm_enum.kOff:
+            if attr_to_update == attribute.PercentSetting:
+                if order == OrderEnum.Ascending:
+                    asserts.assert_equal(attr_to_verify_value_current, init_setting,
+                                         f"[FC] Current {attr_to_verify.__name__} attribute value must be equal to the previous value")
+                elif order == OrderEnum.Descending:
+                    asserts.assert_greater(attr_to_verify_value_current, init_setting,
```
[the remark this answers is not in the document]
Comment (raul-marquez-csa) [reply src/python_testing/TC_FAN_3_1.py:320]: By the time we get here, in the case you're pointing out, which is updating percent descending when initial state is off:
`asserts.assert_greater(attr_to_verify_value_current, init_setting,`

- `init_setting` is 0, it's the value read before testing/writing begins, this in iteration 1 becomes the previous-value.
- `attr_to_verify_value_current` is the value after the first write operation in iteration 1, since it's descending, it began at 100.

Therefore the current value (100) must be greater than previous value (initial value) 0.

File: src/python_testing/TC_FAN_3_1.py [hunk 18]
```diff
@@ -30,140 +30,631 @@
...
+        # Get fan initial state (FanMode, PercentSetting, SpeedSetting if present)
+        init_fan_mode = await self.read_setting(endpoint, attribute.FanMode)
+        init_percent_setting = await self.read_setting(endpoint, attribute.PercentSetting)
+        init_speed_setting = await self.read_setting(endpoint, attribute.SpeedSetting) if self.supports_multispeed else None
+
+        # Determine the initial values of the attributes to update and verify
+        init_attr_to_update_value = init_percent_setting if attr_to_update == attribute.PercentSetting else init_fan_mode
+        init_attr_to_verify_value = init_fan_mode if attr_to_verify == attribute.FanMode else init_percent_setting
+
+        return attr_to_verify, iteration_range, init_fan_mode, init_percent_setting, init_speed_setting, init_attr_to_update_value, init_attr_to_verify_value
```
[the remark this answers is not in the document]
Comment (raul-marquez-csa) [reply src/python_testing/TC_FAN_3_1.py:274]: Done.

File: src/python_testing/TC_FAN_3_1.py [hunk 19]
```diff
@@ -30,140 +30,631 @@
...
+        # Get the current attribute-to-update value from the attribute report queue
+        attr_to_verify_value_current = await self.get_attribute_value_from_queue(attr_to_verify)
+
+        # # Since this was a successful write operation, and the current and previous
+        # # value are the same, it means that no attribute-to-verify value change was
+        # # triggered, setting it to None to skip the verification step
+        # if attr_to_verify_value_current == attr_to_verify_value_previous:
+        #     attr_to_verify_value_current = None
+
+        if attr_to_verify_value_current is not None:
```
[the remark this answers is not in the document]
Comment (raul-marquez-csa) [reply src/python_testing/TC_FAN_3_1.py:546]: This is the change detection mechanism. As discussed previously, I'm not using expectation logic for this due to the need to also handle the **INVALID_IN_STATE** status code.

I cannot know beforehand which write operation will result in an **INVALID_IN_STATE** or **SUCCESS** status code; therefore, I cannot prepare a list of expected values.

The way it's currently written allows me to check in real time which verification to perform without expectations. A **SUCCESS** status code means a greater-than or less-than verification will performed, while an **INVALID_IN_STATE** status code means an equality verification will be performed.

This is unrelated to **Auto** mode.

File: src/python_testing/TC_FAN_3_1.py [hunk 20]
```diff
@@ -21,149 +21,638 @@
...
+        elif self.fan_mode_sequence == 3:
+            fan_modes = [fm_enum.kOff, fm_enum.kLow, fm_enum.kHigh, fm_enum.kAuto]
+        elif self.fan_mode_sequence == 4:
+            fan_modes = [fm_enum.kOff, fm_enum.kHigh, fm_enum.kAuto]
+        elif self.fan_mode_sequence == 5:
+            fan_modes = [fm_enum.kOff, fm_enum.kHigh]
+
+        self.fan_modes = [f for f in fan_modes if not (remove_auto and f == fm_enum.kAuto)]
+
+    async def get_initial_parametes(self, endpoint, attr_to_update, order) -> InitSettings:
```
Comment (cecille) [inline src/python_testing/TC_FAN_3_1.py:222]:
```suggestion
    async def get_initial_parameters(self, endpoint, attr_to_update, order) -> InitSettings:
```

File: src/python_testing/TC_FAN_3_1.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (cecille) [reply src/python_testing/TC_FAN_3_1.py:546]: This function only ever gets called if the response was success.

File: src/python_testing/TC_FAN_3_1.py (hunk 20 above)
Comment (raul-marquez-csa) [reply src/python_testing/TC_FAN_3_1.py:222]: Done

File: src/python_testing/TC_FAN_3_1.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (raul-marquez-csa) [reply src/python_testing/TC_FAN_3_1.py:545]: You're right.
I added context on why this is handled like that. Ignore the invalid in state explanation.

File: src/python_testing/TC_FAN_3_1.py [hunk 21]
```diff
@@ -21,149 +21,245 @@
 # === BEGIN CI TEST ARGUMENTS ===
 # test-runner-runs:
 #   run1:
-#     app: ${ALL_CLUSTERS_APP}
-#     app-args: --discriminator 1234 --KVS kvs1 --trace-to json:${TRACE_APP}.json
+#     app: ${AIR_PURIFIER_APP}
+#     app-args: --discriminator 1234 --KVS kvs1 --trace_file json:${TRACE_APP}.json
```
Comment (cecille) [inline src/python_testing/TC_FAN_3_1.py:25]: Can you please create a follow up issue to get the trace-to option supported on the air purifier on the CI? Not a blocker for TE2, just want to have them all be the same so we can eventually pull out some of this boilerplate

**Review by cecille — APPROVED** [review]

File: src/python_testing/TC_FAN_3_1.py (hunk 21 above)
Comment (raul-marquez-csa) [reply src/python_testing/TC_FAN_3_1.py:25]: will do

File: src/python_testing/TC_FAN_3_1.py [hunk 22]
```diff
@@ -21,149 +21,245 @@
...
+        self.step(3)
+        feature_map = await self.read_setting(attr.FeatureMap)
+        self.supports_multispeed = bool(feature_map & cluster.Bitmaps.Feature.kMultiSpeed)
+
+        # *** NEXT STEPS ***
+        # TH tests the following scenarios and verifies the correct progression of the PercentSetting,
+        # PercentCurrent, FanMode, and if supported, SpeedSetting and SpeedCurrent attributes
+        await self.testing_scenario(attr_to_update=attr.PercentSetting, order=OrderEnum.Ascending)
+        await self.testing_scenario(attr_to_update=attr.PercentSetting, order=OrderEnum.Descending)
+        await self.testing_scenario(attr_to_update=attr.FanMode, order=OrderEnum.Ascending)
```
[the remark this answers is not in the document]
Comment (raul-marquez-csa) [reply src/python_testing/TC_FAN_3_1.py:261]: It currently does that.

[the remark this answers is not in the document]
Comment (cecille) [reply src/python_testing/TC_FAN_3_1.py:261]: `asserts.assert_equal(len(<whatever is in the mode queue>), len(<whatever is in the percent queue>)` after the mode change tests.

[the remark this answers is not in the document]
Comment (raul-marquez-csa) [reply src/python_testing/TC_FAN_3_1.py:279]: updated.

**Review by raju-apple — APPROVED** [review]
